### ISSUE TITLE
[native-image] crashes "host C compiler or linker does not seem to work" with undefined symbols

### ISSUE THREAD

**Issue by aclement** [oracle/graal#1920]
Hi, trying to upgrade my spring app to 19.3 (Java8 flavour). Have i installed it wrong? As the native-image process is about to finish I get:
```
[webflux-netty:22487]        write:   1,076.35 ms
Fatal error: java.lang.RuntimeException: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1

Running command: cc -v -o /Users/aclement/gits/spring-graal-native/spring-graal-native-samples/webflux-netty/unpack/BOOT-INF/classes/webflux-netty -Wl,-no_compact_unwind -Wl,-exported_symbols_list -Wl,/var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415/exported_symbols.list -Wl,-x -arch x86_64 -L/var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415 -L/Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib -L/Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64 /var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415/webflux-netty.o /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libffi.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/liblibchelper.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libjvm.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libstrictmath.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libzip.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libnet.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libjava.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libnio.a -lm -Wl,-framework,CoreFoundation -lpthread -lpthread -ldl -lz

Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/ld" -demangle -lto_library /Library/Developer/CommandLineTools/usr/lib/libLTO.dylib -dynamic -arch x86_64 -macosx_version_min 10.14.0 -o /Users/aclement/gits/spring-graal-native/spring-graal-native-samples/webflux-netty/unpack/BOOT-INF/classes/webflux-netty -L/var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415 -L/Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib -L/Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64 -no_compact_unwind -exported_symbols_list /var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415/exported_symbols.list -x /var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415/webflux-netty.o /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libffi.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/liblibchelper.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libjvm.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libstrictmath.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libzip.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libnet.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libjava.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libnio.a -lm -framework CoreFoundation -lpthread -lpthread -ldl -lz -lSystem /Library/Developer/CommandLineTools/usr/lib/clang/10.0.1/lib/darwin/libclang_rt.osx.a
Undefined symbols for architecture x86_64:
  "_JVM_AssertionStatusDirectives", referenced from:
      _methods in libjava.a(ClassLoader.o)
  "_JVM_DefineClassWithSource", referenced from:
      _Java_java_lang_ClassLoader_defineClass2 in libjava.a(ClassLoader.o)
      _Java_java_lang_ClassLoader_defineClass1 in libjava.a(ClassLoader.o)
  "_JVM_FindClassFromBootLoader", referenced from:
      _Java_java_lang_ClassLoader_findBootstrapClass in libjava.a(ClassLoader.o)
  "_JVM_FindLoadedClass", referenced from:
      _Java_java_lang_ClassLoader_findLoadedClass0 in libjava.a(ClassLoader.o)
  "_JVM_GetResourceLookupCache", referenced from:
      _Java_sun_misc_URLClassPath_getLookupCacheForClassLoader in libjava.a(URLClassPath.o)
  "_JVM_GetResourceLookupCacheURLs", referenced from:
      _Java_sun_misc_URLClassPath_getLookupCacheURLs in libjava.a(URLClassPath.o)
  "_JVM_IsSupportedJNIVersion", referenced from:
      _Java_java_lang_ClassLoader_00024NativeLibrary_load in libjava.a(ClassLoader.o)
  "_JVM_KnownToNotExist", referenced from:
      _Java_sun_misc_URLClassPath_knownToNotExist0 in libjava.a(URLClassPath.o)
  "_JVM_LatestUserDefinedLoader", referenced from:
      _Java_sun_misc_VM_latestUserDefinedLoader0 in libjava.a(VM.o)
  "_JVM_LoadLibrary", referenced from:
      _Java_java_lang_ClassLoader_00024NativeLibrary_load in libjava.a(ClassLoader.o)
  "_JVM_ResolveClass", referenced from:
      _Java_java_lang_ClassLoader_resolveClass0 in libjava.a(ClassLoader.o)
  "_JVM_UnloadLibrary", referenced from:
      _Java_java_lang_ClassLoader_00024NativeLibrary_unload in libjava.a(ClassLoader.o)
      _Java_java_lang_ClassLoader_00024NativeLibrary_load in libjava.a(ClassLoader.o)
  "_Java_sun_misc_Unsafe_defineClass", referenced from:
      ___svm_cglobaldata_base in webflux-netty.o
  "_VerifyClassname", referenced from:
      _Java_sun_misc_URLClassPath_knownToNotExist0 in libjava.a(URLClassPath.o)
      _Java_java_lang_ClassLoader_findBootstrapClass in libjava.a(ClassLoader.o)
  "_VerifyFixClassname", referenced from:
      _Java_sun_misc_URLClassPath_knownToNotExist0 in libjava.a(URLClassPath.o)
      _Java_java_lang_ClassLoader_findBootstrapClass in libjava.a(ClassLoader.o)
      _Java_java_lang_ClassLoader_defineClass2 in libjava.a(ClassLoader.o)
      _Java_java_lang_ClassLoader_defineClass1 in libjava.a(ClassLoader.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
	at java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1005)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:462)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:315)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:454)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:115)
Caused by: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1

Running command: cc -v -o /Users/aclement/gits/spring-graal-native/spring-graal-native-samples/webflux-netty/unpack/BOOT-INF/classes/webflux-netty -Wl,-no_compact_unwind -Wl,-exported_symbols_list -Wl,/var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415/exported_symbols.list -Wl,-x -arch x86_64 -L/var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415 -L/Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib -L/Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64 /var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415/webflux-netty.o /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libffi.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/liblibchelper.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libjvm.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libstrictmath.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libzip.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libnet.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libjava.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libnio.a -lm -Wl,-framework,CoreFoundation -lpthread -lpthread -ldl -lz

Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/ld" -demangle -lto_library /Library/Developer/CommandLineTools/usr/lib/libLTO.dylib -dynamic -arch x86_64 -macosx_version_min 10.14.0 -o /Users/aclement/gits/spring-graal-native/spring-graal-native-samples/webflux-netty/unpack/BOOT-INF/classes/webflux-netty -L/var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415 -L/Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib -L/Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64 -no_compact_unwind -exported_symbols_list /var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415/exported_symbols.list -x /var/folders/tg/klfmdszs2z129xkyrpyzlfyc0000gn/T/SVM-8689304333970909415/webflux-netty.o /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libffi.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/liblibchelper.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libjvm.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/svm/clibraries/darwin-amd64/libstrictmath.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libzip.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libnet.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libjava.a /Users/aclement/installs/graalvm-ce-java8-19.3.0/Contents/Home/jre/lib/libnio.a -lm -framework CoreFoundation -lpthread -lpthread -ldl -lz -lSystem /Library/Developer/CommandLineTools/usr/lib/clang/10.0.1/lib/darwin/libclang_rt.osx.a
Undefined symbols for architecture x86_64:
  "_JVM_AssertionStatusDirectives", referenced from:
      _methods in libjava.a(ClassLoader.o)
  "_JVM_DefineClassWithSource", referenced from:
      _Java_java_lang_ClassLoader_defineClass2 in libjava.a(ClassLoader.o)
      _Java_java_lang_ClassLoader_defineClass1 in libjava.a(ClassLoader.o)
  "_JVM_FindClassFromBootLoader", referenced from:
      _Java_java_lang_ClassLoader_findBootstrapClass in libjava.a(ClassLoader.o)
  "_JVM_FindLoadedClass", referenced from:
      _Java_java_lang_ClassLoader_findLoadedClass0 in libjava.a(ClassLoader.o)
  "_JVM_GetResourceLookupCache", referenced from:
      _Java_sun_misc_URLClassPath_getLookupCacheForClassLoader in libjava.a(URLClassPath.o)
  "_JVM_GetResourceLookupCacheURLs", referenced from:
      _Java_sun_misc_URLClassPath_getLookupCacheURLs in libjava.a(URLClassPath.o)
  "_JVM_IsSupportedJNIVersion", referenced from:
      _Java_java_lang_ClassLoader_00024NativeLibrary_load in libjava.a(ClassLoader.o)
  "_JVM_KnownToNotExist", referenced from:
      _Java_sun_misc_URLClassPath_knownToNotExist0 in libjava.a(URLClassPath.o)
  "_JVM_LatestUserDefinedLoader", referenced from:
      _Java_sun_misc_VM_latestUserDefinedLoader0 in libjava.a(VM.o)
  "_JVM_LoadLibrary", referenced from:
      _Java_java_lang_ClassLoader_00024NativeLibrary_load in libjava.a(ClassLoader.o)
  "_JVM_ResolveClass", referenced from:
      _Java_java_lang_ClassLoader_resolveClass0 in libjava.a(ClassLoader.o)
  "_JVM_UnloadLibrary", referenced from:
      _Java_java_lang_ClassLoader_00024NativeLibrary_unload in libjava.a(ClassLoader.o)
      _Java_java_lang_ClassLoader_00024NativeLibrary_load in libjava.a(ClassLoader.o)
  "_Java_sun_misc_Unsafe_defineClass", referenced from:
      ___svm_cglobaldata_base in webflux-netty.o
  "_VerifyClassname", referenced from:
      _Java_sun_misc_URLClassPath_knownToNotExist0 in libjava.a(URLClassPath.o)
      _Java_java_lang_ClassLoader_findBootstrapClass in libjava.a(ClassLoader.o)
  "_VerifyFixClassname", referenced from:
      _Java_sun_misc_URLClassPath_knownToNotExist0 in libjava.a(URLClassPath.o)
      _Java_java_lang_ClassLoader_findBootstrapClass in libjava.a(ClassLoader.o)
      _Java_java_lang_ClassLoader_defineClass2 in libjava.a(ClassLoader.o)
      _Java_java_lang_ClassLoader_defineClass1 in libjava.a(ClassLoader.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

	at com.oracle.svm.hosted.image.NativeBootImageViaCC.write(NativeBootImageViaCC.java:382)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:652)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Error: Image build request failed with exit status 1
```

**Comment by aclement** [comment]
This is on mac, not yet upgraded to catalina.

**Comment by christianwimmer** [comment]
I can reproduce the `undefined reference to `Java_sun_misc_Unsafe_defineAnonymousClass'` by making `Unsafe.defineAnonymousClass` reachable. The other error are probably because some native methods of `ClassLoader` are seen as reachable.

**Comment by aclement** [comment]
Another flavour of this running my springmvc-tomcat sample:
```
Undefined symbols for architecture x86_64:
  "_Java_sun_misc_Unsafe_defineClass", referenced from:
      ___svm_version_info in tc.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

        at com.oracle.svm.hosted.image.NativeBootImageViaCC.write(NativeBootImageViaCC.java:382)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:652)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)
```

This is using the more recent (9-Dec-2019) hash of graal #ad53b55127ad82 (with openjdk1.8.0_232-jvmci-19.3-b05).

Recreate:
```
git clone git@github.com:spring-projects-experimental/spring-graal-native.git
git checkout upgrade_19_3
cd spring-graal-native/spring-graal-native-feature
mvn clean package
cd ../spring-graal-native-samples/springmvc-tomcat
./compile.sh
```

**Comment by eginez** [comment]
seems related to https://github.com/oracle/graal/issues/1946

**Comment by aclement** [comment]
I think fixed by the #1946 fixes. thanks!

**Comment by michael-simons** [comment]
FYI I cannot compile this https://github.com/michael-simons/neo4j-java-driver-native-example project natively:

```
$GRAALVM_HOME/bin/native-image --version
GraalVM Version 19.3.0 CE
uname -a
Darwin trenzalore 19.0.0 Darwin Kernel Version 19.0.0: Wed Sep 25 20:18:50 PDT 2019; root:xnu-6153.11.26~2/RELEASE_X86_64 x86_64
clang --version
Apple clang version 11.0.0 (clang-1100.0.33.16)
Target: x86_64-apple-darwin19.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
